### ISSUE TITLE
针对训练和推理mask操作的一些疑问

### ISSUE THREAD

**Issue by QingliangMeng** [SWivid/F5-TTS#330]
我在做onnx的导出和尝试训练的时候发现一些问题，虽然我已经解决了一些mask的问题，但是还是想请教一下，我的问题如下：

1、在训练阶段，没有存在text_mask和mel_mask的情况，这样模型不会注意到很多无用的信息么？
例如：
```
        # lens and mask
        if not exists(lens):
            lens = torch.full((batch,), seq_len, device=device)

        mask = lens_to_mask(lens, length=seq_len)  # useless here, as collate_fn will pad to max length in batch

        # get a random span to mask out for training conditionally
        frac_lengths = torch.zeros((batch,), device=self.device).float().uniform_(*self.frac_lengths_mask)
        rand_span_mask = mask_from_frac_lengths(lens, frac_lengths)

        if exists(mask):
            rand_span_mask &= mask
```
注释提到了：useless here, as collate_fn will pad to max length in batch，倘若我传入lens = torch.tensor([245, 356])，inp = torch.rand([2, 500, 100])，代码会报错。所以我认为咱们并没有进行mel_mask。

2、第二段代码是，text_len会padding到seq_len的长度，再去过convnextv2的结构。
```
        batch, text_len = text.shape[0], text.shape[1]
        text = text + 1  # use 0 as filler token. preprocess of batch pad -1, see list_str_to_idx()
        text = text[:, :seq_len]  # curtail if character tokens are more than the mel spec tokens
        text = F.pad(text, (0, seq_len - text_len), value = 0)
```
这段代码我也有一些费解，没有text_mask的遮掩，convnextv2的计算是否正确？因为我在onnx定长方案导出的过程中，明显发现会语意捕捉失败，导致合成失效。我更改增加了许多mask的操作，才使得onnx定长推理成功。

综上：请问作者是如何考量text_mask和mel_mask的问题的？

**Comment by SWivid** [comment]
关于1.：
一方面我们sampler直接从小到大排序了，所以padding得很少，尤其数据量大的时候
另一方面loss计算有个mask，只会算TTS infilling task要重建部分的mse loss

关于2：
这个我觉得是有更好做法的。我们还没有非常深入研究refinement的过程，仅做了model structure search
如果方便的欢迎分享加了怎么样的mask能work的

**Comment by QingliangMeng** [comment]
感谢您的回复。
针对1：确实sort可以减小padding的错误信息，我也会在我的data streaming中进行该操作。在大数据上确实这种误差不是那么重要了。

针对2：其实我为了保证onnx导出没有误差，进行大范围的mask操作，可能会有一些冗余，但是可以指出一些结论，希望可以帮到：
1、主要受到padding信息干扰最严重的位置是：ConvNeXtV2Block。最开始根据理论，我以为self.dwconv会产生较大padding错误干扰，但是后来发现后边的计算也会导致误差错误，所以我基本是一个操作一个mask（这个不严谨，只是为了偷懒争取不错）。
2、Transformer DiT block那个mask确保attention不会造成误差。但是其他DiT Block的算子不知道会不会，为了保险我在每层输出位置进行mask纠正。

**Comment by QingliangMeng** [comment]
十分感谢！确实解决了我很多疑惑哈哈哈。我仔细思考了一下，很赞同这两个猜测（确实有可能）。我会在接下来实验中，操作观测一下。

**Comment by awei669** [comment]
> 感谢您的回复。 针对1：确实sort可以减小padding的错误信息，我也会在我的data streaming中进行该操作。在大数据上确实这种误差不是那么重要了。
> 
> 针对2：其实我为了保证onnx导出没有误差，进行大范围的mask操作，可能会有一些冗余，但是可以指出一些结论，希望可以帮到： 1、主要受到padding信息干扰最严重的位置是：ConvNeXtV2Block。最开始根据理论，我以为self.dwconv会产生较大padding错误干扰，但是后来发现后边的计算也会导致误差错误，所以我基本是一个操作一个mask（这个不严谨，只是为了偷懒争取不错）。 2、Transformer DiT block那个mask确保attention不会造成误差。但是其他DiT Block的算子不知道会不会，为了保险我在每层输出位置进行mask纠正。

请问您是如何添加text_mask的，是直接将filler token部分当成mask吗？

**Comment by QingliangMeng** [comment]
> > 感谢您的回复。 针对1：确实sort可以减小padding的错误信息，我也会在我的data streaming中进行该操作。在大数据上确实这种误差不是那么重要了。
> > 针对2：其实我为了保证onnx导出没有误差，进行大范围的mask操作，可能会有一些冗余，但是可以指出一些结论，希望可以帮到： 1、主要受到padding信息干扰最严重的位置是：ConvNeXtV2Block。最开始根据理论，我以为self.dwconv会产生较大padding错误干扰，但是后来发现后边的计算也会导致误差错误，所以我基本是一个操作一个mask（这个不严谨，只是为了偷懒争取不错）。 2、Transformer DiT block那个mask确保attention不会造成误差。但是其他DiT Block的算子不知道会不会，为了保险我在每层输出位置进行mask纠正。
> 
> 请问您是如何添加text_mask的，是直接将filler token部分当成mask吗？

那不能够。。。。训练阶段我遵循作者讲数据进行sort，减少了无用的padding信息，并且换了显存大的gpu。工程化如果要定长推理的话，我区分了padding token和filler token，filler token不需要mask，是需要将多余适配定长的padding token mask就行了
